bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/PageImpl.java

```diff
@@ -0,0 +1,451 @@
+        }
+
+        com.day.cq.wcm.api.Page page = getRootPage();
+
+        if (page != null) {
+            return getModelUrl(request, page);
+        }
+
+        return null;
+    }
+
+    /**
+     * @return Returns the root model of the given page
+     */
+    @Nullable
+    @Override
+    public Page getRootModel() {
+        if (currentStyle != null && currentStyle.containsKey(PR_IS_ROOT)) {
+            return this;
+        }
+
+        com.day.cq.wcm.api.Page rootPage = getRootPage();
+
+        if (rootPage == null) {
+            return null;
+        }
+
+        return modelFactory.getModelFromWrappedRequest(
+            PageImpl.getHierarchyServletRequest(request, rootPage, currentPage),
+            rootPage.getContentResource(),
+            this.getClass());
+    }
+
+    /**
+     * Returns a model URL for the given page URL
+     *
+     * @param url page URL
+     * @return {@link String} model URL
+     */
+    protected static String getModelUrl(@Nonnull String url) {
+        if (StringUtils.isBlank(url)) {
+            return null;
+        }
+
+        int dotIndex = url.indexOf('.');
+
+        if (dotIndex < 0) {
+            dotIndex = url.length();
+        }
+
+        return url.substring(0, dotIndex) + URL_MODEL_EXTENSION;
+    }
+
+    /**
+     * Wrap the provided request to ensure the static references to the containing page of a component is accurate
+     *
+     * @param request request to be wrapped
+     * @param page page to be referenced as statically containing the current page content
+     * @return {@link SlingHttpServletRequest} a {@link SlingHttpServletRequestWrapper} containing given page and request
+     */
+    protected static SlingHttpServletRequest getHierarchyServletRequest(@Nonnull SlingHttpServletRequest request,
+                                                                        @Nonnull com.day.cq.wcm.api.Page page,
+                                                                        @Nullable com.day.cq.wcm.api.Page entryPage) {
+        // Request attribute key of the component context
+        final String ATTR_COMPONENT_CONTEXT = "com.day.cq.wcm.componentcontext";
+
+        // Request attribute key of the current page
+        final String ATTR_CURRENT_PAGE = "currentPage";
+
+        SlingHttpServletRequest wrapper = new SlingHttpServletRequestWrapper(request);
+
+        ComponentContext componentContext = (ComponentContext) request.getAttribute(ATTR_COMPONENT_CONTEXT);
+
+        // When traversing child pages the currentPage must be updated
+        wrapper.setAttribute(ATTR_COMPONENT_CONTEXT, new HierarchyComponentContextWrapper(componentContext, page));
+        wrapper.setAttribute(ATTR_CURRENT_PAGE, page);
+        wrapper.setAttribute(ATTR_HIERARCHY_ENTRY_POINT_PAGE, entryPage);
+
+        return wrapper;
+    }
+
+
+    // style helpers zone
+
+    /**
+     * Returns the first numeric selector. The default value is 0
```

Comment:
Could you please clarify this javadocs?
E.g.: Returns the tree depth that is configured in the policy.